File: src/claw_mem/storage/episodic.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Episodic Memory Storage

Stores daily conversation records, format: memory/YYYY-MM-DD.md
"""

import os
from datetime import datetime, timedelta
from typing import List, Dict, Optional
from pathlib import Path


class EpisodicStorage:
    """
    Episodic Memory Storage
    
    Storage location: ~/.openclaw/workspace/memory/YYYY-MM-DD.md
    
    Features:
    - Date-based file storage
    - Auto-expiry (default 30 days)
    - Markdown format, human-readable
    """
    
    def __init__(self, workspace: Path, ttl_days: int = 30):
        """
        Initialize Episodic Storage
        
        Args:
            workspace: Workspace path
            ttl_days: Memory retention days
        """
        self.workspace = workspace
        self.memory_dir = workspace / "memory"
        self.ttl_days = ttl_days
        
        # Ensure directory exists
        self.memory_dir.mkdir(parents=True, exist_ok=True)
    
    def store(self, memory_record: Dict) -> None:
        """
        Store memory to today's file
        
        Args:
            memory_record: Memory record
        """
        # Get today's filename
        today = datetime.now().strftime("%Y-%m-%d")
        file_path = self.memory_dir / f"{today}.md"
        
        # Format memory content
        content = self._format_memory(memory_record)
        
        # Append to file
        with open(file_path, "a", encoding="utf-8") as f:
            f.write(content + "\n")
    
    def get_all(self) -> List[Dict]:
        """
        Get all Episodic memories
        
        Returns:
            List[Dict]: Memory records
        """
        return self._read_all_files()
    
    def get_recent(self, limit: int = 20) -> List[Dict]:
        """
        Get recent memories
        
        Args:
            limit: Number of results
            
        Returns:
            List[Dict]: Memory records
        """
        memories = []
        
        # Get recent N files
        files = sorted(self.memory_dir.glob("*.md"), reverse=True)[:limit]
        
        for file_path in files:
            memories.extend(self._read_file(file_path))
        
        return memories[:limit]
    
    def get_by_date(self, date: str) -> List[Dict]:
        """
        Get memories by date
        
        Args:
            date: Date string (YYYY-MM-DD)
            
        Returns:
            List[Dict]: Memory records
        """
        file_path = self.memory_dir / f"{date}.md"
        if file_path.exists():
            return self._read_file(file_path)
        return []
    
    def cleanup_expired(self) -> int:
        """
        Clean up expired memories
        
        Returns:
            int: Number of deleted files
        """
        cutoff_date = datetime.now() - timedelta(days=self.ttl_days)
        deleted_count = 0
        
        for file_path in self.memory_dir.glob("*.md"):
            # Parse date from filename
            try:
                file_date = datetime.strptime(file_path.stem, "%Y-%m-%d")
                if file_date < cutoff_date:
                    file_path.unlink()
                    deleted_count += 1
            except ValueError:
                continue
        
        return deleted_count
    
    def count(self) -> int:
        """
        Get number of memory records
        
        Returns:
            int: Record count
        """
        count = 0
        for file_path in self.memory_dir.glob("*.md"):
            count += len(self._read_file(file_path))
        return count
    
    def _format_memory(self, memory_record: Dict) -> str:
        """
        Format memory record to Markdown
        
        Args:
            memory_record: Memory record
            
        Returns:
            str: Markdown formatted content
        """
        timestamp = memory_record.get("timestamp", datetime.now().isoformat())
        content = memory_record.get("content", "")
        tags = memory_record.get("tags", [])
        metadata = memory_record.get("metadata", {})
        
        # Add metadata comments
        meta = []
        if tags:
            meta.append(f"tags: {', '.join(tags)}")
        if memory_record.get("session_id"):
            meta.append(f"session: {memory_record['session_id']}")
        
        # Add custom metadata fields
        for key, value in metadata.items():
            meta.append(f"{key}: {value}")
        
        # Format output
        lines = []
        if meta:
            lines.append("<!-- " + "; ".join(meta) + " -->")
        lines.append(f"[{timestamp}] {content}")
        
        return "\n".join(lines)
    
    def _read_all_files(self) -> List[Dict]:
        """
        Read all memory files
        
        Returns:
            List[Dict]: All memory records
        """
        memories = []
        
        for file_path in sorted(self.memory_dir.glob("*.md"), reverse=True):
            memories.extend(self._read_file(file_path))
        
        return memories
    
    def _read_file(self, file_path: Path) -> List[Dict]:
        """
        Read memory file
        
        Args:
            file_path: File path
            
        Returns:
            List[Dict]: Memory records
        """
        memories = []
        
        if not file_path.exists():
            return memories
        
        with open(file_path, "r", encoding="utf-8") as f:
            current_meta = {}
            for line in f:
                line = line.strip()
                if not line:
                    continue
                
                # Parse metadata comments
                if line.startswith("<!--") and line.endswith("-->"):
                    meta_content = line[4:-3].strip()
                    for item in meta_content.split(";"):
                        if ":" in item:
                            key, value = item.split(":", 1)
                            current_meta[key.strip()] = value.strip()
                # Parse memory content
                elif line.startswith("["):
                    try:
                        end_timestamp = line.index("]")
                        timestamp = line[1:end_timestamp]
                        content = line[end_timestamp+1:].strip()
                        
                        # Extract standard fields
                        session_id = current_meta.get("session")
                        tags_str = current_meta.get("tags", "")
                        
                        # Extract custom metadata (exclude standard fields)
                        metadata = {}
                        for key, value in current_meta.items():
                            if key not in ["tags", "session"]:
                                metadata[key] = value
                        
                        memories.append({
                            "timestamp": timestamp,
                            "content": content,
                            "tags": tags_str.split(", ") if tags_str else [],
                            "session_id": session_id,
                            "metadata": metadata,
                            "type": "episodic",
                            "source": str(file_path)
                        })
                        current_meta = {}  # Reset metadata
                    except (ValueError, IndexError):
                        continue
        
        return memories
    
    def __repr__(self) -> str:
        return f"EpisodicStorage(dir={self.memory_dir}, ttl={self.ttl_days}d)"

File: src/storage/episodic.ts
import type { EpisodicEntry } from "../types.js";
import * as fs from "fs";
import * as path from "path";

/**
 * Episodic Memory Storage.
 *
 * Storage: `memory/YYYY-MM-DD.md` — one Markdown file per day.
 * Markdown format with HTML comment metadata blocks.
 *
 * Behaviour matches Python EpisodicStorage exactly.
 */
export class EpisodicStorage {
  private memoryDir: string;
  private ttlDays: number;

  constructor(workspace: string, ttlDays: number = 30) {
    // Use workspace directly as memory storage directory
    this.memoryDir = workspace;
    this.ttlDays = ttlDays;
    fs.mkdirSync(this.memoryDir, { recursive: true });
  }

  /** Store a memory record to today's file (atomic write via temp + rename). */
  store(record: Record<string, unknown>): void {
    const today = new Date().toISOString().slice(0, 10);
    const filePath = path.join(this.memoryDir, `${today}.md`);
    const newContent = this.formatMemory(record) + "\n";
    const existing = fs.existsSync(filePath) ? fs.readFileSync(filePath, "utf-8") : "";
    const tmpPath = filePath + ".tmp." + Date.now();
    fs.writeFileSync(tmpPath, existing + newContent, "utf-8");
    fs.renameSync(tmpPath, filePath);
  }

  /** Batch store: append multiple records in a single file write. */
  storeBatch(records: Array<Record<string, unknown>>): void {
    if (records.length === 0) return;
    const today = new Date().toISOString().slice(0, 10);
    const filePath = path.join(this.memoryDir, `${today}.md`);
    const existing = fs.existsSync(filePath) ? fs.readFileSync(filePath, "utf-8") : "";
    const newContent = records.map((r) => this.formatMemory(r)).join("\n") + "\n";
    const tmpPath = filePath + ".tmp." + Date.now();
    fs.writeFileSync(tmpPath, existing + newContent, "utf-8");
    fs.renameSync(tmpPath, filePath);
  }

  /** Get all episodic memories across all files. */
  getAll(): EpisodicEntry[] {
    return this.readAllFiles();
  }

  /**
   * Get the most recent N entries from the newest date files.
   * Python equivalent: get_recent(limit) — reads newest files, returns up to limit entries.
   */
  getRecent(limit: number = 20): EpisodicEntry[] {
    const files = this.globMdFiles().sort().reverse();
    const memories: EpisodicEntry[] = [];
    for (const fp of files) {
      const filePath = path.join(this.memoryDir, fp);
      const entries = this.readFile(filePath);
      for (let i = entries.length - 1; i >= 0; i--) {
        memories.push(entries[i]);
        if (memories.length >= limit) return memories;
      }
    }
    return memories;
  }

  /** Get entries for a specific date (YYYY-MM-DD). */
  getByDate(date: string): EpisodicEntry[] {
    const filePath = path.join(this.memoryDir, `${date}.md`);
    return this.readFile(filePath);
  }

  /** Clean up files older than ttlDays. Returns count of deleted files. */
  cleanupExpired(): number {
    const cutoff = new Date();
    cutoff.setDate(cutoff.getDate() - this.ttlDays);
    let deleted = 0;
    for (const fname of this.globMdFiles()) {
      try {
        const dateStr = fname.replace(".md", "");
        const [y, m, d] = dateStr.split("-").map(Number);
        const fileDate = new Date(y, m - 1, d);
        if (fileDate < cutoff) {
          fs.unlinkSync(path.join(this.memoryDir, fname));
          deleted++;
        }
      } catch { continue; }
    }
    return deleted;
  }

  /** Count total memory records. */
  count(): number {
    let total = 0;
    for (const fname of this.globMdFiles()) {
      total += this.readFile(path.join(this.memoryDir, fname)).length;
    }
    return total;
  }

  // ── format ──────────────────────────────────────────────────────

  private formatMemory(record: Record<string, unknown>): string {
    const timestamp = record.timestamp || new Date().toISOString();
    const content = record.content || "";
    const tags = (record.tags as string[]) || [];
    const metadata = (record.metadata as Record<string, unknown>) || {};
    const sessionId = record.session_id as string | undefined;
    const meta: string[] = [];
    if (tags.length) meta.push(`tags: ${tags.join(", ")}`);
    if (sessionId) meta.push(`session: ${sessionId}`);
    for (const [k, v] of Object.entries(metadata)) {
      meta.push(`${k}: ${v}`);
    }
    let lines = "";
    if (meta.length) lines += `<!-- ${meta.join("; ")} -->\n`;
    lines += `[${timestamp}] ${content}`;
    return lines;
  }

  // ── file IO ─────────────────────────────────────────────────────

  private globMdFiles(): string[] {
    try {
      return fs.readdirSync(this.memoryDir).filter((f) => f.endsWith(".md"));
    } catch { return []; }
  }

  private readAllFiles(): EpisodicEntry[] {
    const memories: EpisodicEntry[] = [];
    for (const fname of this.globMdFiles().sort().reverse()) {
      memories.push(...this.readFile(path.join(this.memoryDir, fname)));
    }
    return memories;
  }

  private readFile(filePath: string): EpisodicEntry[] {
    if (!fs.existsSync(filePath)) return [];
    const entries: EpisodicEntry[] = [];
    let currentMeta: Record<string, string> = {};
    try {
      const lines = fs.readFileSync(filePath, "utf-8").split("\n");
      for (const raw of lines) {
        const line = raw.trim();
        if (!line) continue;
        if (line.startsWith("<!--") && line.endsWith("-->")) {
          const metaContent = line.slice(4, -3).trim();
          for (const item of metaContent.split(";")) {
            const trimmed = item.trim();
            const idx = trimmed.indexOf(":");
            if (idx !== -1) {
              const key = trimmed.slice(0, idx).trim();
              const val = trimmed.slice(idx + 1).trim();
              currentMeta[key] = val;
            }
          }
          continue;
        }
        if (line.startsWith("[")) {
          const endIdx = line.indexOf("]");
          if (endIdx !== -1) {
            const timestamp = line.slice(1, endIdx);
            const content = line.slice(endIdx + 1).trim();
            const sessionId = currentMeta["session"];
            const tagsStr = currentMeta["tags"] || "";
            const tags = tagsStr ? tagsStr.split(", ").filter(Boolean) : [];
            const metadata: Record<string, string> = {};
            for (const [k, v] of Object.entries(currentMeta)) {
              if (k !== "tags" && k !== "session") metadata[k] = v;
            }
            entries.push({
              timestamp,
              content,
              tags,
              session_id: sessionId,
              id: metadata["id"] || undefined,
              metadata,
              type: "episodic",
              source: filePath,
            });
            currentMeta = {};
          }
        }
      }
    } catch { /* file read error → empty */ }
    return entries;
  }
}
